xpotato/dataset/dataset.py

```diff
@@ -1,35 +1,109 @@
 
 
 class Dataset:
     def __init__(
-        self, examples: List[Tuple[str, str]], label_vocab: Dict[str, int], lang="en"
+        self,
+        examples: List[Tuple[str, str]] = None,
+        label_vocab: Dict[str, int] = {},
+        lang="en",
+        path=None,
+        binary=False,
     ) -> None:
         self.label_vocab = label_vocab
-        self._dataset = self.read_dataset(examples)
+        if path:
+            self._dataset = self.read_dataset(path=path, binary=binary)
+        else:
+            self._dataset = self.read_dataset(examples=examples)
         self.extractor = GraphExtractor(lang=lang, cache_fn=f"{lang}_nlp_cache")
         self.graphs = None
 
-    def read_dataset(self, examples: List[Tuple[str, str]]) -> List[Sample]:
-        return [Sample(example) for example in examples]
+    @staticmethod
+    def save_dataframe(df: pd.DataFrame, path: str) -> None:
+        graphs = [graph_to_pn(graph) for graph in df["graph"].tolist()]
+        df["graph"] = graphs
+        df.to_csv(path, index=False, sep="\t")
+
+    def prune_graphs(self, graphs: List[nx.DiGraph] = None) -> None:
+        graphs_str = []
+        for i, graph in enumerate(graphs):
+            graph.remove_nodes_from(list(nx.isolates(graph)))
+            # ADAM: THIS IS JUST FOR PICKLE TO PENMAN CONVERSION
+            graph = self._random_postprocess(graph)
+
+            g = [
+                c
+                for c in sorted(
+                    nx.weakly_connected_components(graph), key=len, reverse=True
+                )
+            ]
+            if len(g) > 1:
+                print(
+                    "WARNING: graph has multiple connected components, taking the largest"
+                )
+                g_pn = graph_to_pn(graph.subgraph(g[0].copy()))
+            else:
+                g_pn = graph_to_pn(graph)
+
+            graphs_str.append(g_pn)
+
+        return graphs_str
+
+    def read_dataset(
+        self,
+        examples: List[Tuple[str, str]] = None,
+        path: str = None,
+        binary: bool = False,
+    ) -> List[Sample]:
+        if examples:
+            return [Sample(example, PotatoGraph()) for example in examples]
+        elif path:
+            if binary:
+                df = pd.read_pickle(path)
+                graphs_str = self.prune_graphs(df.graph.tolist())
+                df.drop(columns=["graph"], inplace=True)
+                df["graph"] = graphs_str
+            else:
+                df = pd.read_csv(path, sep="\t")
+
+            return [
+                Sample(
+                    (example["text"], example["label"]),
+                    potato_graph=PotatoGraph(graph_str=example["graph"]),
+                    label_id=example["label_id"],
+                )
+                for _, example in tqdm(df.iterrows())
+            ]
+        else:
+            raise ValueError("No examples or path provided")
+
+    # ADAM: THIS WILL NEED TO BE ADDRESSED
```

Comment:
what is this about? Maybe open an issue and reference it in the comment

Reply:
Did in #41